linkerd/app/outbound/src/lib.rs

```diff
@@ -448,22 +458,54 @@ impl Config {
             .into_inner()
             .into_make_service();
 
-        // The stack is served lazily since caching layers spawn tasks from
-        // their constructor. This helps to ensure that tasks are spawned on the
-        // same runtime as the proxy.
-        // Forwards TCP streams that cannot be decoded as HTTP.
-        let tcp_forward = svc::stack(tcp::Forward::new(tcp_connect))
+        // Load balances TCP streams that cannot be decoded as HTTP.
+        let tcp_balance = svc::stack(tcp_connect.clone())
+            .push_make_thunk()
             .push(admit::AdmitLayer::new(prevent_loop))
-            .push_map_target(TcpEndpoint::from);
+            .check_make_service::<TcpEndpoint, ()>()
+            .push(discover::resolve(map_endpoint::Resolve::new(
+                endpoint::FromMetadata,
+                resolve,
+            )))
+            .push(discover::buffer(1_000, cache_max_idle_age))
+            .push_map_target(Addr::from)
+            .push_on_response(tcp::balance::layer(EWMA_DEFAULT_RTT, EWMA_DECAY))
+            .push_fallback_with_predicate(
+                svc::stack(tcp_connect.clone())
+                    .push_make_thunk()
+                    .push(admit::AdmitLayer::new(prevent_loop))
+                    .push_map_target(TcpEndpoint::from),
+                is_discovery_rejected,
+            )
```

Comment:
nit/tioli: might be nice to add `trace` or `debug` spans in here indicating whether we are in the tcp balancer or forwarding to the original dst?